### ISSUE TITLE
[feature request] RL algorithms supporting custom recurrent policies

### ISSUE THREAD

**Issue by AdamGleave** [hill-a/stable-baselines#241]
Many RL algorithms handle recurrent and non-recurrent policies differently. Currently, this switch in behavior is achieved by testing whether the policy class is a subclass of `LstmPolicy`. 

Examples:
```
./stable_baselines/a2c/a2c.py:101:                if issubclass(self.policy, LstmPolicy):                                                                                                             
./stable_baselines/ppo1/pposgd_simple.py:249:                                      shuffle=not issubclass(self.policy, LstmPolicy))                                                                   
./stable_baselines/acktr/acktr_disc.py:118:                if issubclass(self.policy, LstmPolicy):                                                                                                    
./stable_baselines/her/her.py:88:            assert not issubclass(self.policy, LstmPolicy), "Error: cannot use a recurrent policy for the HER model."                                                
./stable_baselines/acer/acer_simple.py:183:                if issubclass(self.policy, LstmPolicy):                                                                                                    
./stable_baselines/ppo2/ppo2.py:113:                if issubclass(self.policy, LstmPolicy):                                                                                                           
./stable_baselines/common/base_class.py:98:            assert not issubclass(self.policy, LstmPolicy) or self.n_envs == env.num_envs, \    
```

All the recurrent policies implemented in Stable Baselines inherit from `LstmPolicy`, so this works OK for typical use cases. However, some users may want to define their own recurrent policies. Right now, we are forcing them to subclass from `LstmPolicy`, which is fairly restrictive.

I can see two main alternatives:
  - Add a new abstract class, `RecurrentPolicy`, which subclasses `BasePolicy` but does not define any methods. Switch all subclass checks to `RecurrentPolicy`. `LstmPolicy` inherits from `RecurrentPolicy`.
  - Add a class attribute, `recurrent`, and check for whether this is true or false rather than using `issubclass`.

Any preferences between those two, or any other ideas? I'm happy to work on a PR for this once  a design is agreed.

**Comment by araffin** [comment]
Pinging @erniejunior for that conversation, as he has been working on recurrent policies (see #158 and #161 recurrent policies with keras).

**Comment by erniejunior** [comment]
Sorry for answering so slowly! Thanks @AdamGleave for opening the issue. I agree with you that the way we handle recurrent policies does need some rework:

Distinguishing between recurrent and non-recurrent policies is important. What sets them apart from other policies is not their recurrence but rather that they have some kind of *state*. Therefore I would vote for distinguishing between stateless and stateful (is that a word?) policies.

I have no strong opinion on the flag vs. subclass decision but I would lean slightly more towards option 2 (which you seem to already have implemented). What do you think about the following idea?: make the `BasePolicy` stateful by default. A stateless policy would then just return and accept `None` states. A rough sketch would be:
```python
class BasePolicy:
    ...
    @property
    def initial_state():
        return #None if stateless, numpy array if stateful with fixed initial state, tf.Variable if learned initial state

    def is_stateless():
        return self.initial_state is None

    def step(observation, state=None):
        # Stateless policies just ignore the state argument
```
Actually that the way we have it implemented is already quite close to this: `BasePolicy` has the state as an argument to the step method. For some reason the initial state is a member of `ActorCriticPolicy`. We need to move this up to the `BasePolicy` and make it a property (to make clear that it is not writable).

While we are at it we should clean up (or at least properly document) the rest of the policy interface as well. Looking through the code of PPO2 I found that it accesses member variables of the policy such as `states_ph`, `masks_ph`, `proba_step`, `value` and `initial_state`. I think we have not documented what those members are supposed to mean/contain and I also think that we should turn them into **documented** properties of `BasePolicy` or `ActorCriticPolicy` (e.g. it took me a while to figure out what the `masks_ph` actually does because it is badly named and not documented).

When I played around with the LSTM policies the results were mediocre at best and I dropped them in favor of linear policies with an observation history. Maybe we should **introduce some tests** to check if recurrent policies really do perform better than linear policies. Is there some simple benchmark environment that requires learning n-step dependencies? This would not just check if recurrence is better but also if our implementation is even correct at all.

Before you start (re)implementing LSTMs be aware that right now they are a mess in tensorflow (see my issue here tensorflow/tensorflow#24850) and maybe it is worth to wait for the 2.0 API. I found that execution speed varies widely which we also should keep in mind!

**Comment by araffin** [comment]
Maybe related: https://github.com/openai/baselines/pull/859

**Comment by AdamGleave** [comment]
@erniejunior thanks for this comment. I agree with distinguishing based on statefulness rather than recurrence; I'll rename the `recurrent` attribute to `stateful`. Also agree with suggested changes to `BasePolicy`. At a conference now so may not have time to work on this until later in the week, but will try and mock up some changes.

I'm not sure if I'll have bandwidth to make other changes to the policy interface, but I'll try and clear things up a bit.

Thanks for the helpful notes on LSTM polices. In my case I'm finetuning some existing policies (which are LSTM policies), which is why I wanted a custom policy class, but I'll bear this in mind when I have a choice over architecture. 

Not aware off the top of my head of any standard benchmarks that require arbitrary n-step dependencies. Maybe maze navigation tasks? If you slowed down the agent / made the corridors longer, then it would require a larger observation history, but wouldn't need more memory / hidden state.

**Comment by araffin** [comment]
> Maybe we should introduce some tests to check if recurrent policies really do perform better than linear policies. Is there some simple benchmark environment that requires learning n-step dependencies?
> Not aware off the top of my head of any standard benchmarks that require arbitrary n-step dependencies. 

For those kind of benchmark, I've got some for you ;)

First for debugging, I would suggest removing velocity from observation in environment like `CartPole` or `MountainCar` (and check that a feedforward net is not working without frame-stacking).
Then, for assessing the performance, I would go for Atari games when applicable (start with the simple `Pong`) without frame-stacking.

**Comment by araffin** [comment]
>Therefore I would vote for distinguishing between stateless and stateful (is that a word?) policies.

I find the name `stateful` a bit confusing, as you can have `stateless` LSTM... (see [here](http://philipperemy.github.io/keras-stateful-lstm/) or [here](https://datascience.stackexchange.com/questions/26663/when-to-use-stateful-lstm).
What was wrong with the word `recurrent`?

Pinging @hill-a for that kind of API big choices.

**Comment by erniejunior** [comment]
I was confused by `stateful` too at first. But apparently it is a common term in computer science. I would be OK to switch it to `stateless` and invert the logic if you think we should stick to more common English words.

When I understand the stateless keras LSTMs correctly they would be exactly the reason why we should distinguish between stateful/stateless instead of recurrent/non-recurrent. There can be recurrent networks that have no state and I also could think of policies with state that are not recurrent (e.g. when you want to do frame stacking in your policy instead of the environment).

**Comment by hill-a** [comment]
I really do not like `stateful` or `stateless` as names, why not just call them what they are: `state policy` and `base policy`?

As for the solution, the OOP dev in me is not happy, but I still strongly agree with the proposed solution. of defining a default `initial_state() -> None` function that needs to be overrided for correct usage.

**Comment by erniejunior** [comment]
> I really do not like `stateful` or `stateless` as names, why not just call them what they are: `state policy` and `base policy`?

Any particular reason you do not like `stateful` and `stateless`? I am not particularly fond of them either but by naming them this way we make very clear that the state in this context refers to the internal state of the policy and *not* the system state of our MDP. Not knowing about policies with internal state I might interpret `StatePolicy` as a policy that acts directly on environment states instead of just observations. What do you think about `uses_internal_state` or `has_internal_state` or simply `has_state`?

> As for the solution, the OOP dev in me is not happy, but I still strongly agree with the proposed solution. of defining a default `initial_state() -> None` function that needs to be overrided for correct usage.

I feel with you here ;-)
I had the following use case in mind: what if I write some code that has to deal with a number of policies. Some are with state. Some are stateless. The only way to not have to deal with stateful policies as an edge case is to just treat all policies as stateful -- some just always stay in the same `None` state. The only place left where we have to distinguish between stateful and stateless policies is during training because here we need to arrange the batches differently.

**Comment by AdamGleave** [comment]
I don't have a strong opinion on whether to call it `stateful` or `recurrent`, happy to go with whatever you all prefer.

I agree from an OO perspective the class structure doesn't make much sense. A stateless class being a degenerate case of a stateful class seems OK though.

**Comment by araffin** [comment]
@AdamGleave @hill-a @erniejunior 
It seems we need to vote otherwise, it won't advance...
If there is a tie, either we hope people from the community to help us decide, otherwise we can let @hill-a have the final word as owner of this repo.

Then (correct me if I'm wrong), the candidates are:
- recurrent
- stateful
- other?

I vote for recurrent.

**Comment by AdamGleave** [comment]
My vote is for `recurrent`, since when I hear `state` in the context of RL I tend to think of the MDP state. However, I'd change my mind if there was an example of a policy that needed state that was not recurrent.

**Comment by erniejunior** [comment]
My vote is for `stateful` or `has_memory` or `has_state`. With the reasoning that 
1. There [somehow seem to be stateless LSTM](https://github.com/hill-a/stable-baselines/issues/241#issuecomment-477536699).
2. A policy does not have to be a neural network. It could be any program that computes an action from an observation. If that program uses persistent memory instead of just acting like a (mathematical) function we would say that is has a memory or it is based on a state. Only in the (arguably currently dominating) field of neural networks this circumstance is referred to as recurrence. By using the term `recurrent` we implicitly assume that all policies are neural networks. At least we should be aware of this.
A practical example of a non-recurrent NN policy with state would be a policy that bases its action prediction on the last n obersvations. It could keep those last n observations in its state. Right now I use dirty hacks where I modify or wrap an environment to do the observation stacking for me.
3. `recurrent` refers to the internal structure of the policy which is not what matters in our context. The fact that we need to pass a state is the important thing here.

**Comment by AdamGleave** [comment]
So far recurrent has two votes and stateful one vote. @hill-a do you want to weigh in / cast the deciding vote? I'd like to get this PR merged soon :)

**Comment by hill-a** [comment]
Sorry, I was trying to understand stateless LSTMs before voting. It seems that stateless LSTMs have an internal state 

(https://datascience.stackexchange.com/questions/36440/whats-the-difference-of-stateless-lstm-and-a-normal-feed-forward-nn) 

It's just not preserved between batches. 

(https://github.com/keras-team/keras/blob/master/keras/layers/recurrent.py#L2121 and https://github.com/keras-team/keras/blob/master/keras/engine/base_layer.py#L46) 

In our case stateless LSTMs dont make that much sense I think. So I would prefere sticking to `recurrent`, as the only way to achieve recurrency would be with an internal state with Reinforcement learning.

**Comment by araffin** [comment]
>There somehow seem to be stateless LSTM.

as @hill-a stated, those have still an internal state that is dropped between batches.

>A policy does not have to be a neural network.

On that point, I totally agree, you can parametrize your policy in many different ways. However, in the current state of the lib, we have only neural network policies and provide only recurrent policies (LSTM here) that derive from the base class.

>At least we should be aware of this.
>NN policy with state would be a policy that bases its action prediction on the last n obersvations. It could keep those last n observations in its state.
>The fact that we need to pass a state is the important thing here.

Good point @erniejunior , then we should state explicitly in the docstring of the class what it can do, even so the current lib only provide recurrent policies. 

> recurrent refers to the internal structure of the policy which is not what matters in our context.

I disagree with that last point, as "recurrent" can also refer to [recurrence relation](https://en.wikipedia.org/wiki/Recurrence_relation) which just says that it needs previous terms in order to compute the current one.

**Comment by araffin** [comment]
closed by #244 being merged
